### PR TITLE
CODEC-291: Hex Encode/Decode with existing arrays

### PR DESCRIPTION
**Pull request by adamretter**
Additional encode/decode methods for the Hex class where you can pass in the array to write the data to.
This allows you to work with existing arrays. This is useful for both reducing allocations and prepending/appending to existing data.

### CODE REVIEW

File: src/main/java/org/apache/commons/codec/binary/Hex.java [hunk 1]
```diff
@@ -148,12 +169,62 @@
...
+     *
+     * @param data a byte[] to convert to Hex characters
+     * @param dataOffset the position in {@code data} to start encoding from
+     * @param dataLen the number of bytes from {@code dataOffset} to encode
+     * @param toLowerCase {@code true} converts to lowercase, {@code false} to uppercase
+     * @param out a char[] which will hold the resultant appropriate characters from the alphabet.
+     * @param outOffset the position within {@code out} at which to start writing the encoded characters.
+     * @since 1.15
+     */
+    protected static void encodeHex(final byte[] data, final int dataOffset, final int dataLen,
```
Comment (aherbert) [inline src/main/java/org/apache/commons/codec/binary/Hex.java:205]: Should be public

File: src/main/java/org/apache/commons/codec/binary/Hex.java [hunk 2]
```diff
@@ -148,12 +169,62 @@
...
+     *
+     * @param data a byte[] to convert to Hex characters
+     * @param dataOffset the position in {@code data} to start encoding from
+     * @param dataLen the number of bytes from {@code dataOffset} to encode
+     * @param toLowerCase {@code true} converts to lowercase, {@code false} to uppercase
+     * @return A char[] containing the appropriate characters from the alphabet For best results, this should be either
+     *         upper- or lower-case hex.
+     * @since 1.15
+     */
+    protected static char[] encodeHex(final byte[] data, final int dataOffset, final int dataLen,
```
Comment (aherbert) [inline src/main/java/org/apache/commons/codec/binary/Hex.java:187]: Should be public

**Comment by adamretter** [comment]
@aherbert Thanks. I have rebased, I will try and add the missing test coverage in the next few days...

**Comment by adamretter** [comment]
@aherbert Okay I just added the missing tests.

**Comment by aherbert** [comment]
Squashed and merged to master

**Pull request closed without merging**